### ISSUE TITLE
TilizeWithValPadding support height sharding

### ISSUE THREAD

**Issue by nsmithtt** [tenstorrent-metal/tt-metal#5965]
Currently `TilizeWithValPadding` only supports width sharding, convs are often height sharded so they're unable to use `TilizeWithValPadding` without going to interleaved which is expensive.

**Comment by nsmithtt** [comment]
This would be nice to have for unet bringup, but I've marked as only a P2 because I think we have a way to work around it, prototyping that today.  Will escalate this issue if need be.

**Comment by jliangTT** [comment]
Hi @nsmithtt , does this fall into the tensor manipulation bucket or in the conv op bucket? (trying to make sure these operators are triaged correctly)

**Comment by nsmithtt** [comment]
@jliangTT, tensor manipulation

**Comment by prajaramanTT** [comment]
@nsmithtt Can we consider this ticket done ? If yes, can you please close this ? Thanks

**Comment by ntarafdar** [comment]
This can be generalized to support sharding , use sharded address generator , this is a reference PR where we add sharded address generator to an exiting op.
https://github.com/tenstorrent/tt-metal/pull/17692

**Comment by ntarafdar** [comment]
Also add this for TIlize (without val padding)

**Comment by cmaryanTT** [comment]
Putting this up for bounty. This should be solvable with insertion of the sharded address generator.

**Comment by CheeksTheGeek** [comment]
hi, super interested in this!

Was wondering about a few things:
- should this task include pure height sharding, or when you say height, you also mean that it should support compound i.e. H&W or block sharding (haven't looked into whether the latter is somewhere alr in the codebase)
- in a height shard, should the pad value cover only the halo rows that spill past each shard' local height or should it try to replicate the full pad pattern so each shard is self contained.

- w.r.t to the codebase structure, it should make sense to replicate how [`ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/height_sharded_reader_common.hpp`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/conv/conv2d/device/kernels/height_sharded_reader_common.hpp) was put in place, just to confirm?

**Comment by cguerreroTT** [comment]
@cmaryanTT or @nsmithtt can you please take a look at @CheeksTheGeek's questions above?

**Comment by ntarafdar** [comment]
> * should this task include pure height sharding, or when you say height, you also mean that it should support compound i.e. H&W or block sharding (haven't looked into whether the latter is somewhere alr in the codebase)
> *

Hello @CheeksTheGeek I think the initial ask was for the input tensor to be height sharded and thus the output tensor would be height sharded as well.

However if you use the sharded_addr_gen , this is an in-place replacement for the interleaved-add gen any kernel that supports interleaved should support all types of sharding for free.

This is a sample PR that added sharded addr gen to an existing kernel.

You'll want to at the top op level see if its interleaved use the current path, if its width sharded use the current path, 
if its any other sharded type use your modified interleaved kernel that now has the sharded addr gen instead, and you'll need to modify the host and kernel to send the extra sharding information through.

**Comment by ntarafdar** [comment]
@CheeksTheGeek 
This PR might even be more helpful as we did something like this for untilize
https://github.com/tenstorrent/tt-metal/pull/22735
https://github.com/tenstorrent/tt-metal/pull/23198

**Comment by CheeksTheGeek** [comment]
Hi @ntarafdar, thanks for the guidance. Based on those PRs and your guidance, this is the rough plan I'm htinking of right now for this. Please let me know if this aligns with your expectations or if I should adjust any part of it.  Excited to take this task on! (cc: @Shubhamsaboo)
1. Single Core Path:
    - I'll update the validate() and the compute_output_specs() to accept height-shared inputs Specifically, if the memory layout is HEIGHT_SHARED, we can allow it and propgagate that to the output spec, where each shard will have the padded height.
    - I'll make use of the ShardedAddrGen in the device kernel, whereby the writer kernel will generate the output tile addresses based on the height-shard index. Essentially, I'll adapt the existing tilize writer kernel so that it fills tiles by rows instead of columns (or maybe uses SHardedAddr with something like an added split_dim=Height). This could handle inserting the pad_value at the correct edges for each shard.
    - A thing I'll make sure of is that each shard only pads its local boundary rows (so that resembled shards match the global tilized tensor). And if instead the shard needs to produce the full pad pattern internally, I think there should be an explicit clarification that the assumption is only for local padding.
2. Multicore Path
    - I'll modify the program factory to detect the same HEIGHT_SHARDED strategy, and we can add a new variant (example tilize_with_val_padding_multi_core_row_interleaved)analogous to the existing col_interleaved variant.
        - In this variant, assign each core a range of rows (using grid_to_cores with appropriate ordering), and launch the tilize kernel on each core with that range. This will leverage the same ShardedAddrGen logic as a single core, but each core will process its subset of tiles in parallel.
    - We can ensure that any existing interleaved or width sharded branches remain unchanged, so we only add new branches and not break the older ones. What I mean by this is, for example if the input is interleaved or width-sharded, we use the current paths, but if the input is height sharded, we use the new path. This is pretty obvious, but had to document here just for reminding myself later on.
3. For the actual kernel impl:
    - I'll most likely need to rework/add a writer kernel under `tilize_with_val_padding/device/kernels`, and if needed, something like a `writer_unary_stick_layout_split_rows.cpp` where the  kernel will loop over the tile coordinates and use either ShardedAddrGen or a manual row-based offset computation.

Just wanted to confirm a few things:
- for the pad distribution, should each height shard core output include the full pad pattern or only pad at its local edges. I assume local edges only, but please confirm.
- for the mixed sharding: if in the future someone uses block-sharding, do we need to support that in this PR or is it a separate, large enough task beyond the scope of this one.

**Comment by cguerreroTT** [comment]
Hey @CheeksTheGeek I've assigned this bounty to you.

Be sure that you read over the [BOUNTY TERMS](https://docs.tenstorrent.com/bounty_terms.html), as you agree to all of them by picking up this issue.

If we haven't received any communication from you in a two week period, we will re-assign the issue, and you will no longer be eligible for completion of this bounty.

If you have any questions feel free to ping someone here on this issue. I'll let @ntarafdar confirm your plan above.  Look forward to seeing your submission!

**Comment by cguerreroTT** [comment]
Hi @ntarafdar, just following up on the questions above.  Can you please take a look at @CheeksTheGeek's plan?

@CheeksTheGeek have you started working on this bounty already?

**Comment by CheeksTheGeek** [comment]
Yes, I started running some tests and stuff, but wanted to follow up on the questions above :)

**Comment by ntarafdar** [comment]
hey @CheeksTheGeek , sorry for the delay the height shard should only include the padding it has. 

in terms of block sharding, I think if you use sharded_addr_gen it should implicitly include block sharding support

**Comment by ntarafdar** [comment]
@CheeksTheGeek how's this going? is there anything else you need?

**Comment by CheeksTheGeek** [comment]
Hi @ntarafdar, really sorry for the late reply.
Wanted to give a few updates: Because of some changes to Koyeb's usage model + exam season at university right now, my work on this task got slowed down but I'm getting back to speed and I should have good news in a few days.
I think atleast for the design portion and your rec about sharded_addr_gen, I've gotten that part figured out, will make a PR soon.

**Comment by marty1885** [comment]
@CheeksTheGeek Hope you are doing well. Anything we can do on our end?

**Comment by marty1885** [comment]
@CheeksTheGeek Last call. I'll close and reassign the bounty if you are not interested (and I will assume you are not interested anymore if no reply is made).

**Comment by CheeksTheGeek** [comment]
Hey @marty1885  ,  Huge apologies for my severe lack of communication with this.

I spent a few weeks going back to this, I thought I understood how to approach this, but I think I still need to dig more w.r.t some details, one of the biggest blocker was having severe problems with Koyeb (not running it but rather the slowness of the iterations from making a change to pushing it to a repo and using it as a container build source, with building the repo and running the pytests on the instances). Also, their support has been terrible. I'll give it another try this weekend.

Best,
Chaitanya

**Comment by ntarafdar** [comment]
Hey @CheeksTheGeek wanted to see if this was looked at. We might want to prioritize this ourselves soon, but if you have already solved it I don't want to take it away from you.

**Comment by marty1885** [comment]
@CheeksTheGeek any luck and are you still interested in the bounty?

**Comment by marty1885** [comment]
Hello <username>, any progress made? I am obligated to check in periodically and reassign stale bounties. Let me know if you are making progress/still interested or not. Else I'll re-open it by Monday

**Comment by Mountagha** [comment]
@marty1885 hellow.

**Comment by marty1885** [comment]
Thanks @Mountagha it's yours now, thank you for the interest! Please let us know within 2 weeks your progress. Please refer to our [Bounty Terms & Conditions](https://docs.tenstorrent.com/bounty_terms.html). Good luck with the project! And let us know if there's any bug/issues on our side that is stopping you find making the bounty happen. 

Note to future self - assigned @Mountagha 2 bounties right now. But it's the holidays so it'll let him have more to play with.

**Comment by Mountagha** [comment]
CC @ntarafdar can I DM you on discord regarding this task ? Thanks

**Comment by jberkowitzTT** [comment]
@Mountagha - do you still need help on the next steps for this issue?

**Comment by jberkowitzTT** [comment]
Feel free to ask questions on this thread & we can get you the support you need!

**Comment by Mountagha** [comment]
@jberkowitzTT, nop found what I was looking for in the official documentation. I will ask here next time. Thanks

**Comment by jberkowitzTT** [comment]
Sounds good! Do you have an estimate for when a PR will be ready?

**Comment by Mountagha** [comment]
> Sounds good! Do you have an estimate for when a PR will be ready?

A week (Next friday) would be a good estimate. If it's too long and it's top priority I might squeeze it to early next week like monday or tuesday.

**Comment by Mountagha** [comment]
> [@CheeksTheGeek](https://github.com/CheeksTheGeek) This PR might even be more helpful as we did something like this for untilize [#22735](https://github.com/tenstorrent/tt-metal/pull/22735) [#23198](https://github.com/tenstorrent/tt-metal/pull/23198)

Also was wondering as mentionned in this PR is it better to go single core first (1 PR ) and then move to multi core after the former is validated. ?

**Comment by jberkowitzTT** [comment]
@nsmithtt - do you have any thoughts on the above?

**Comment by Mountagha** [comment]
> [@nsmithtt](https://github.com/nsmithtt) - do you have any thoughts on the above?

CC @nsmithtt , help make my weekend busy [joke] 😎 😄

**Comment by nsmithtt** [comment]
Going to defer to @ntarafdar, I don't have much context around this op.  I would assume that single core is a good first step, but for this issue to be fully resolved it requires multi-core support.

**Comment by jberkowitzTT** [comment]
Hi @Mountagha, thanks for the hard work on this!

What are the next steps here? Is a part 2 needed before we can close the part 1 PR?

**Comment by Mountagha** [comment]
> Hi @Mountagha, thanks for the hard work on this!
> 
> What are the next steps here? Is a part 2 needed before we can close the part 1 PR?

Hello @jberkowitzTT yess I'm debugging some issues before opening the second PR. Expect it tomorrow or at most Friday evening. 
Thanks

**Comment by lmee** [comment]
Implementation approach (claim plan): 1) Reproduce issue with deterministic steps; 2) implement minimal fix; 3) add tests and concise PR notes.

**Comment by lmee** [comment]
Implementation approach (claim plan): 1) Reproduce issue with deterministic steps; 2) implement minimal fix; 3) add tests and concise PR notes.

**Comment by lmee** [comment]
Implementation approach (claim plan): 1) Reproduce issue with deterministic steps; 2) implement minimal fix; 3) add tests and concise PR notes.

**Comment by lmee** [comment]
I went through "[Bounty $1500] TilizeWithValPadding support height sharding" and reviewed the complete issue discussion. The core problem appears to be: Currently only supports width sharding, convs are often height sharded so they're unable to use without going to interleaved which is expensive. My root-cause hypothesis is missing conversion/lowering coverage for the target op path, plus incomplete edge-case validation. A key detail from the discussion is: "should this task include pure height sharding, or when you say height, you also mean that it should support compound i.e. H&W or block sharding (haven't looked into whether the lat." My plan is to build a deterministic failing case in the current conversion/lowering pipeline, implement the missing op mapping and validation path, then add IR/kernel regression coverage with a concise perf/compat note. I can start now and keep updates focused on evidence and validation. Best, Jerry Li

**Comment by jberkowitzTT** [comment]
Hi @Mountagha - the work you've done satisfies the completion requirements so we can close this out & start the payment process

Congratulations & thank you for the great work!
